staging/vhost-device-gpu/src/vhu_gpu.rs

```diff
@@ -811,42 +831,148 @@ mod tests {
+            .unwrap();
+        VirtioGpu {
+            rutabaga,
+            resources: BTreeMap::default(),
+            fence_state: Arc::new(Mutex::new(Default::default())),
+        }
+    }
+
     #[test]
-    fn process_requests_no_desc() {
-        let (mut backend, _, vring) = init();
+    fn test_process_queue_chain() {
+        let (mut backend, mem, _) = init();
+
+        backend.update_memory(mem.clone()).unwrap();
 
-        // Descriptor chain size zero, shouldn't fail
-        // backend
-        //     .process_requests(Vec::<GpuDescriptorChain>::new(), &vring)
-        //     .unwrap();
+        let vring = VringRwLock::new(mem.clone(), 0x1000).unwrap();
+        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+        vring.set_queue_ready(true);
+
+        let mut buf: Vec<u8> = vec![0; 30];
+        let command_types = [
+            VIRTIO_GPU_CMD_RESOURCE_CREATE_2D,
+            VIRTIO_GPU_CMD_RESOURCE_UNREF,
+            VIRTIO_GPU_CMD_SET_SCANOUT,
+            VIRTIO_GPU_CMD_SET_SCANOUT_BLOB,
+            VIRTIO_GPU_CMD_RESOURCE_FLUSH,
+            VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D,
+            VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING,
+            VIRTIO_GPU_CMD_RESOURCE_DETACH_BACKING,
+            VIRTIO_GPU_CMD_GET_CAPSET,
+            VIRTIO_GPU_CMD_GET_CAPSET_INFO,
+            VIRTIO_GPU_CMD_GET_EDID,
+            VIRTIO_GPU_CMD_CTX_CREATE,
+            VIRTIO_GPU_CMD_CTX_DESTROY,
+            VIRTIO_GPU_CMD_CTX_ATTACH_RESOURCE,
+            VIRTIO_GPU_CMD_CTX_DETACH_RESOURCE,
+            VIRTIO_GPU_CMD_RESOURCE_CREATE_3D,
+            VIRTIO_GPU_CMD_TRANSFER_TO_HOST_3D,
+            VIRTIO_GPU_CMD_TRANSFER_FROM_HOST_3D,
+            VIRTIO_GPU_CMD_SUBMIT_3D,
+            VIRTIO_GPU_CMD_RESOURCE_CREATE_BLOB,
+            VIRTIO_GPU_CMD_RESOURCE_MAP_BLOB,
+            VIRTIO_GPU_CMD_RESOURCE_UNMAP_BLOB,
+            VIRTIO_GPU_CMD_UPDATE_CURSOR,
+            VIRTIO_GPU_CMD_MOVE_CURSOR,
+            VIRTIO_GPU_CMD_RESOURCE_ASSIGN_UUID,
+        ];
+        for cmd_type in command_types {
+            let desc_chain = prepare_desc_chains(&mem, &mut buf, cmd_type);
+            let mem = mem.memory().into_inner();
+
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(Error::CreateReader)
+                .unwrap();
+            let mut writer = desc_chain
+                .clone()
+                .writer(&mem)
+                .map_err(Error::CreateWriter)
+                .unwrap();
+
+            let mut virtio_gpu = new_2d();
+            let mut signal_used_queue = true;
+
+            backend
+                .process_queue_chain(
+                    &mut virtio_gpu,
+                    &mem,
+                    &vring,
+                    desc_chain.head_index(),
+                    &mut reader,
+                    &mut writer,
+                    &mut signal_used_queue,
+                )
+                .unwrap();
```

Comment:
After calling this function I would expect hope for some asserts. But this might require refactoring some of the code to be more testable. Right now it only tests that it doesn't return an error. It doesn't test if some expected behavior has happened.

Reply:
in testing that it doesn't return an error it does test code regions for some codes in `fn process_queue_chain()`, which is important. I can add test to test other regions that returns and err using unwrap_err() variant.

Reply:
Yes it tests that it doesn't return an error. But it doesn't really test that process_queue_chain works, it only technically increases coverage. Now testing that is not easy. I would imagine we would put a command in the descriptor chain, have a mock gpu that returns a response. And check if the response is written to the descriptor chain.

Reply:
```
But it doesn't really test that process_queue_chain works,
```
 I understand this as you referring to a functionality test not a unit test.

```
I would imagine we would put a command in the descriptor chain, have a mock gpu that returns a response. And check if the response is written to the descriptor chain.
```
the purpose of these test is for code coverage, so performing unwraps on some of those functions is to allow the test go through those code lines. by calling unwrap some code inside the function are executed.

Reply:
The definition on Wikipedia for unit test is this:
> Unit testing, a.k.a. component or module testing, is a form of [software testing](https://en.wikipedia.org/wiki/Software_testing) by which isolated [source code](https://en.wikipedia.org/wiki/Source_code) is tested to validate expected behavior.[[1]](https://en.wikipedia.org/wiki/Unit_testing#cite_note-kolawa-1)

Now the part *validate expected behavior* is the part I feel like this test doesn't fully fulfill.

Reply:
the expected result of this function is that it returns `ok` or an `err` variant. I believe that the unwrap is useful because I expect the function to always succeed in this scenario. would it then be fine with you to `assert!(backend.process_queue_chain.is_ok());`.

when you say validate the behaviour can you give an example of what the test would look like then. also this is not the first test written in this manner, you can make references to how other test in the rust-vmm crate are being handled.

Reply:
> when you say validate the expected behaviour can you give an example of what the test would look like then. also this is not the first test written in this manner, you can make references to how other test in the rust-vmm crate are being handled.

I haven't looked into this yet, but sure. If the mainters are fine with this, than sure. (I personally would rather have lower coverage and better tests... But I am not the mainter, so it's not up to me)

The `.unwrap()`  you have is preferable to the  `assert!(backend.process_queue_chain.is_ok());.`

For a proper unit test (like I was taught at university) I would not use the real `VirtioGpu` here (I would test that separately - the `isolated source code` in the definition).  I would call `process_queue_chain` with some command in the descriptor chain. I can then assert that the correct method (with correct arguments) is called in the fake/mock `VirtioGpu`. From the fake `VirtioGpu` I can return a response and assert that `process_queue_chain` wrote the response to the descriptor chain and set the `signal_used_queue` to `true`.

Reply:
```
The .unwrap() you have is preferable to the assert!(backend.process_queue_chain.is_ok());.
```
I thought so too.

```
 I would call process_queue_chain with some command in the descriptor chain. I can then assert that the correct method (with correct arguments) is called in the fake/mock VirtioGpu. From the fake VirtioGpu I can return a response and later assert that process_queue_chain wrote the response to the descriptor chain and set the signal_used_queue to true.
```
If you read the code you will see that I call the process_queue_chain with each of the gpu command in the descriptor chain. I don't know what you define as real or fake, but from what I understand your concern is that there are no explicit assertions in this test. I would like to point out that this test rely on the absence of panics to determine if successful or not (.unwrap). also feel free to add more test that do assertions.

Reply:
By `fake` and `mock` I mean another implementation of `VirtioGpu` used just for testing. 
https://www.softwaretestingmagazine.com/knowledge/unit-testing-fakes-mocks-and-stubs/

By "real" GPU I mean the GPU implementation we currently have, the `RutabagaVirtioGpu` struct. 

> If you read the code you will see that I call the process_queue_chain with each of the gpu command in the descriptor chain. 
Yes, you do this part, I don't have an issue with this. Like you point out I don't like the lack of assertions though.

> I would like to point out that this test rely on the absence of panics to determine if successful or not (.unwrap). also feel free to add more test that do assertions.

Yes, I understand what you are doing. I still feel like this is not a proper unit test without assertions - this test would pass for this function: 
```
 fn process_queue_chain(
        &mut self,
        virtio_gpu: &mut RutabagaVirtioGpu,
        mem: &GuestMemoryMmap,
        vring: &VringRwLock,
        head_index: u16,
        reader: &mut Reader,
        writer: &mut Writer,
        signal_used_queue: &mut bool,
    ) -> Result<()> {
      // maybe some simple if statements if you add tests to check Err return value.
      // notice I don't have to do anything with the `reader` or `writer` or the `virtio_gpu` for your test to pass! 
      Ok(())
}
```
You don't check if the `process_queue_chain` reads the command, if it tries to execute the command using `RutabagaVirtioGpu`, if it writes the response of the command.... 
Effectively the `.unwrap()`  only checks the correctness of the return statment(s).

We can merge it like this for now, and improve the tests later or leave them like this.

Reply:
we should indeed consider adding assertions to verify whether the function correctly reads commands.  I agree we proceed with the merge and plan to enhance the tests later.